### PR TITLE
Deprecate `LossRecorder`, create `LossOutput` dataclass alternative

### PR DESCRIPTION
**Pull request by adamgayoso**
Dataclasses allow the same workflow to be applied to jax-based models.

Two changes are important:

- `LossRecorder` is deprecated, to be removed in v0.20.0
- `JaxTrainingPlan` gets proper logging using the same paradigm

TODO:

- [x] Change each model to use the new `LossOutput`
- [x] Merge #1743 

Some code was added to loss recorder to make it backwards compatible, but it will be removed anyway

### CODE REVIEW

File: scvi/train/_trainingplans.py [hunk 1]
```diff
@@ -298,14 +301,19 @@ def compute_and_log_metrics(
...
+            loss_output = loss_recorder
+        rec_loss = loss_output.reconstruction_loss_sum
+        n_obs_minibatch = loss_output.n_obs_minibatch
+        kl_local = loss_output.kl_local_sum
+        kl_global = loss_output.kl_global_sum
+
+        # Use the torchmetric object for the ELBO
+        # We only need to update the ELBO metric
+        # As it's defined as a sum of the other metrics
+        next(iter(metrics.values()))[0].update(
```
Comment (justjhong) [inline scvi/train/_trainingplans.py:316]: why can't you key into elbo here?

File: scvi/train/_trainingplans.py [hunk 2]
```diff
@@ -702,15 +708,15 @@ def training_step(self, batch, batch_idx, optimizer_idx=0):
             labelled_tensors=labelled_dataset,
         )
         input_kwargs.update(self.loss_kwargs)
-        _, _, scvi_losses = self.forward(full_dataset, loss_kwargs=input_kwargs)
-        loss = scvi_losses.loss
+        _, _, loss_output = self.forward(full_dataset, loss_kwargs=input_kwargs)
+        loss = loss_output.loss
```
Comment (justjhong) [inline scvi/train/_trainingplans.py:712]: is this backwards compatible?

File: scvi/train/_trainingplans.py (hunk 1 above)
Comment (adamgayoso) [reply scvi/train/_trainingplans.py:316]: ah yeah I didn't see we have access to the mode here

File: scvi/train/_trainingplans.py [hunk 3]
```diff
@@ -281,7 +284,7 @@ def forward(self, *args, **kwargs):
     @torch.inference_mode()
     def compute_and_log_metrics(
         self,
-        loss_recorder: LossRecorder,
+        loss_recorder: Union[LossRecorder, LossOutput],
         metrics: MetricCollection,
```
[the remark this answers is not in the document]
Comment (justjhong) [reply scvi/train/_trainingplans.py:288]: commenting on the metrics not the loss recorder

File: scvi/train/_trainingplans.py (hunk 2 above)
Comment (adamgayoso) [reply scvi/train/_trainingplans.py:712]: Yes? I'm just making the variable naming more consistent

File: scvi/module/base/_base_module.py [hunk 4]
```diff
@@ -58,56 +63,147 @@ def __init__(
...
+        return self._loss_output.dict_sum(self._loss_output.reconstruction_loss)
+
+    @property
+    def _reconstruction_loss(self):
+        return self._loss_output.reconstruction_loss
 
     @property
     def kl_local(self) -> Union[torch.Tensor, jnp.ndarray]:  # noqa: D102
-        return self._get_dict_sum(self._kl_local)
+        return self._loss_output.dict_sum(self._loss_output.kl_local)
```
Comment (martinkim0) [inline scvi/module/base/_base_module.py:96]:
```suggestion
        return self.dict_sum(self._loss_output.kl_local)
```

File: scvi/module/base/_base_module.py [hunk 5]
```diff
@@ -58,56 +63,147 @@ def __init__(
...
+        return self._loss_output.kl_local_sum
+
+    @property
+    def kl_global_sum(self) -> Union[torch.Tensor, jnp.ndarray]:  # noqa: D102
+        return self._loss_output.kl_global_sum
 
     @property
     def kl_global(self) -> Union[torch.Tensor, jnp.ndarray]:  # noqa: D102
-        return self._get_dict_sum(self._kl_global)
+        return self._loss_output.dict_sum(self._loss_output.kl_global)
```
Comment (martinkim0) [inline scvi/module/base/_base_module.py:116]:
```suggestion
        return self.dict_sum(self._loss_output.kl_global)
```

File: scvi/module/base/_base_module.py [hunk 6]
```diff
@@ -58,56 +63,147 @@ def __init__(
...
+    This class provides an organized way to record the model loss, as well as
+    the components of the ELBO. This may also be used in MLE, MAP, EM methods.
+    The loss is used for backpropagation during inference. The other parameters
+    are used for logging/early stopping during inference.
+
+    Parameters
+    ----------
+    loss
+        Tensor with loss for minibatch. Should be one dimensional with one value.
+        Note that loss should be in an array/tensory and not a float.
```
Comment (martinkim0) [inline scvi/module/base/_base_module.py:137]:
```suggestion
        Note that loss should be in an array/tensor and not a float.
```

File: scvi/module/base/_base_module.py [hunk 7]
```diff
@@ -58,56 +63,147 @@ def __init__(
...
+    reconstruction_loss
+        Reconstruction loss for each observation in the minibatch. If a tensor, converted to
+        a dictionary with key "reconstruction_loss" and value as tensor
```
Comment (martinkim0) [inline scvi/module/base/_base_module.py:140]:
```suggestion
        a dictionary with key "reconstruction_loss" and value as tensor.
```

File: scvi/module/base/_base_module.py [hunk 8]
```diff
@@ -58,56 +63,147 @@ def __init__(
...
+    kl_local
+        KL divergence associated with each observation in the minibatch. If a tensor, converted to
+        a dictionary with key "kl_local" and value as tensor
```
Comment (martinkim0) [inline scvi/module/base/_base_module.py:143]:
```suggestion
        a dictionary with key "kl_local" and value as tensor.
```

File: scvi/module/base/_base_module.py [hunk 9]
```diff
@@ -58,56 +63,147 @@ def __init__(
...
+    extra_metrics
+        Additional metrics can be passed as arrays/tensors or dictionaries of
+        arrays/tensors.
+    """
+
+    loss: LossRecord
+    reconstruction_loss: Optional[LossRecord] = None
+    kl_local: Optional[LossRecord] = None
+    kl_global: Optional[LossRecord] = None
+    extra_metrics: Optional[Dict[str, Tensor]] = field(default_factory=dict)
```
Comment (martinkim0) [inline scvi/module/base/_base_module.py:156]:
```suggestion
    extra_metrics: Dict[str, Tensor] = field(default_factory=dict)
```

File: scvi/train/_trainingplans.py (hunk 3 above)
Comment (martinkim0) [inline scvi/train/_trainingplans.py:288]:
```suggestion
        metrics: OrderedDict,
```

File: scvi/module/base/_base_module.py [hunk 10]
```diff
@@ -58,56 +63,147 @@ def __init__(
...
-
     @property
     def loss(self) -> Union[torch.Tensor, jnp.ndarray]:  # noqa: D102
-        return self._get_dict_sum(self._loss)
+        return self._loss_output.loss
 
     @property
     def reconstruction_loss(self) -> Union[torch.Tensor, jnp.ndarray]:  # noqa: D102
-        return self._get_dict_sum(self._reconstruction_loss)
+        return self._loss_output.dict_sum(self._loss_output.reconstruction_loss)
```
Comment (martinkim0) [inline scvi/module/base/_base_module.py:88]:
```suggestion
        return self.dict_sum(self._loss_output.reconstruction_loss)
```

**Review by martinkim0 — APPROVED** [review]

File: scvi/module/base/_base_module.py (hunk 9 above)
Comment (adamgayoso) [reply scvi/module/base/_base_module.py:167]: I would say this is optional because you don't need to specifiy it
